core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java

```diff
@@ -1066,7 +1066,119 @@ public ServerBuilder serviceUnder(String pathPrefix, HttpService service) {
         requireNonNull(pathPrefix, "pathPrefix");
         requireNonNull(service, "service");
         warnIfServiceHasMultipleRoutes(pathPrefix, service);
-        return service(Route.builder().pathPrefix(pathPrefix).build(), service);
+        return route().addRoute(Route.builder().pathPrefix(pathPrefix).build()).build(service);
```

Comment:
How about trying to unwrap into an `HttpServiceWithRoutes` and handle multiple routes here? Then we don't need `warnIfServiceHasMultipleRoutes` either, which is a nice improvement. It will also allow us to remove the below method that requires an `HttpServiceWithRoutes`.

Reply:
That's a good suggestion. 👍